src/backends/perf_pt/decode.c

```
@@ -35,13 +35,317 @@
+        *addr = 0;
+        return true;
+    } else if (*decoder_status > 0) {
+        // An unhandled case.
+        return false;
+    }
+
+    // Now fetch the block information.
+    struct pt_block block;
+    *decoder_status = pt_blk_next(decoder, &block, sizeof(block));
+    // Other +ve decoder status codes can arise here. We ignore them for now,
+    // and let them be detected by drain_events() above when we are next
+    // called.
+    if (*decoder_status < 0) {
+        if (*decoder_status == -pte_eos) {
+            // End of stream is flagged as an error in the case of pt_blk_next().
+            *addr = 0;
+            return true;
+        } else {
+            // A real error.
+            return false;
+        }
+    }
+
+    if (block.truncated == 1) {
+        // XXX A truncated block. This occurs when a block straddles a section
+        // boundary. In this case we may need some extra logic, but this should
+        // be a rare case.
+        return false;
+    }
+
+    if (block.ninsn == 0) {
+        // A block with zero instructions. Shouldn't happen.
+        return false;
+    }
+
+    *addr = block.ip;
+    return true;
+}
+
+/*
+ * Given a decoder and the last decoder status, drains events in the PT packet
+ * stream.
+ *
+ * The new decoder status is returned.
+ */
+static int
+drain_events(struct pt_block_decoder *decoder, int decoder_status) {
+    while(decoder_status & pts_event_pending) {
+        struct pt_event event;
+        decoder_status = pt_blk_event(decoder, &event, sizeof(event));
+        if (decoder_status < 0) {
+            return decoder_status;
+        }
+    }
+    return decoder_status;
+}
+
+/*
+ * Loads the libipt image `image` with the code of the current process.
+ *
+ * Returns true on success or false otherwise.
+ */
+static bool
+load_self_image(struct load_self_image_args *args)
+{
+    bool ret = dl_iterate_phdr(load_self_image_cb, args) == 0;
+    if (fsync(args->vdso_fd) == -1) {
```

Can't we simplify this to `return fsync(...) != -1;`?

---

We can!

---

Wait, what if `dl_iterate_phdr` fails, but fsync succeeds? We would return `true` ignoring the error.

---

That's a good point. But should we do `fsync` if `dl_iterate_phdr` failed? Maybe we should really be doing something like:

```
return !((dl_iterate_phdr(load_self_image_cb, args) == 0) && (fsync(args->vdso_fd) != -1));
```

---

We can do, but after splitting the long line, it's probably not going to be pretty. I'll try it nonetheless.

---

Maybe just two `if`s then.

---

It didn't look too bad in the end.